core/src/main/java/cz/o2/proxima/scheme/SchemaDescriptors.java

```diff
@@ -0,0 +1,885 @@
+      }
+      return fieldsByName;
+    }
+
+    public boolean hasField(String name) {
+      return getFieldsAsMap().containsKey(name);
+    }
+
+    @SuppressWarnings("squid:S1452")
+    public SchemaDescriptors.FieldTypeDescriptor<?> getField(String name) {
+      return Optional.ofNullable(getFieldsAsMap().getOrDefault(name, null))
+          .orElseThrow(
+              () ->
+                  new IllegalArgumentException(
+                      "Field " + name + " not found in structure " + getName()));
+    }
+
+    @Nullable
+    @Override
+    public T getDefaultValue() {
+      return defaultValueFactory != null ? defaultValueFactory.apply() : null;
+    }
+
+    @Override
+    public SchemaTypeDescriptor<T> toTypeDescriptor() {
+      return SchemaTypeDescriptor.<T>builder()
+          .type(getType())
+          .structureTypeDescriptor(this)
+          .build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      if (!super.equals(o)) {
+        return false;
+      }
+      @SuppressWarnings("unchecked")
+      StructureTypeDescriptor<?> that = (StructureTypeDescriptor<?>) o;
+      return name.equals(that.name) && fields.equals(that.fields);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), name, fields);
+    }
+  }
+
+  /**
+   * Enum type descriptor.
+   *
+   * @param <T> value type
+   */
+  public static class EnumTypeDescriptor<T extends Serializable> extends GenericTypeDescriptor<T>
+      implements WithDefaultValue<T> {
+
+    @Getter private final List<T> values;
+    @Nullable private final Factory<T> defaultValueFactory;
+
+    public EnumTypeDescriptor(List<T> values, @Nullable Factory<T> defaultValueFactory) {
+      super(AttributeValueType.ENUM);
+      this.values = values;
+      this.defaultValueFactory = defaultValueFactory;
+    }
+
+    @Nullable
+    @Override
+    public T getDefaultValue() {
+      return defaultValueFactory != null ? defaultValueFactory.apply() : null;
+    }
+
+    @Override
+    public SchemaTypeDescriptor<T> toTypeDescriptor() {
+      return SchemaTypeDescriptor.<T>builder().type(getType()).enumTypeDescriptor(this).build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      if (!super.equals(o)) {
+        return false;
+      }
+      @SuppressWarnings("unchecked")
+      EnumTypeDescriptor<?> that = (EnumTypeDescriptor<?>) o;
+      return values.equals(that.values);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), values);
+    }
+  }
+
+  @ToString
+  public static class FieldTypeDescriptor<T> extends GenericTypeDescriptor<T>
```

Comment:
This looks odd to me. All other `*TypeDescriptor`s seem to describe a type. Field is not a type, it is a property of a structure. It *has* a name and a type. I think it could be removed and the name be held in the StructureTypeDescriptor only (it actually already contains the map, the name is redundant there currently).

Reply:
This can be probably replaced by field `Map<String, TypeDescriptor>` and then we can maybe remove it. WDYT?

Reply:
Removed `FieldTypeDescriptor` and refactored.

Reply:
:+1: